File: src/main/java/no/nav/aap/api/sts/SystemToken.java
package no.nav.aap.api.sts;

import com.fasterxml.jackson.annotation.JsonProperty;
import no.nav.security.token.support.core.jwt.JwtToken;

import java.time.Duration;
import java.time.LocalDateTime;

import static no.nav.aap.api.util.StringUtil.limit;
import static no.nav.aap.api.util.TimeUtil.localDateTime;

public record SystemToken(@JsonProperty("access_token") JwtToken accessToken,
                          @JsonProperty("expires_in") Long expiresIn,
                          @JsonProperty("token_type") String tokenType,
                          @JsonProperty("scope") String scope) {

    boolean isExpired(Duration slack) {
        return LocalDateTime.now().isAfter(getExpiration().minus(slack));
    }

    String getToken() {
        return accessToken().getTokenAsString();
    }

    LocalDateTime getExpiration() {
        return localDateTime(accessToken().getJwtTokenClaims().getExpirationTime());
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + "[accessToken=" + limit(accessToken.getTokenAsString(), 12)
                + ", expires=" + getExpiration()
                + ", scope=" + scope
                + ", tokenType=" + tokenType + "]";
    }

}

File: src/main/kotlin/no/nav/aap/api/rest/sts/SystemToken.kt
package no.nav.aap.api.rest.sts

import com.fasterxml.jackson.annotation.JsonProperty
import no.nav.aap.api.util.TimeUtil.localDateTime
import no.nav.security.token.support.core.jwt.JwtToken
import java.time.Duration
import java.time.LocalDateTime.now

data class SystemToken(@JsonProperty("access_token") private val accessToken: JwtToken,
                       @JsonProperty("expires_in") private val expiresIn: Long,
                       @JsonProperty("token_type") private val  tokenType: String,
                       @JsonProperty("scope") private val scope: String) {

    fun isExpired(slack: Duration) =  now().isAfter(expiration?.minus(slack)) ?: true
    val token = accessToken.tokenAsString
    val expiration = localDateTime(accessToken.jwtTokenClaims.expirationTime)

}